src/mlpack/methods/lmnn/lmnn_function.hpp

```diff
@@ -217,6 +218,12 @@ class LMNNFunction
   Constraints<MetricType> constraint;
   //! Holds pre-calculated cij.
   arma::mat pCij;
+  //! Holds the norm of each data point.
+  arma::vec norm;
+  //! Hold previous eval value.
+  double evalOld;
```

Comment:
This confuses me a little bit---shouldn't we cache one `evalOld` for each query point / impostor combination?  I don't think it should just be one `eval` for the whole dataset.

Reply:
Right, It seems like a big mistake. I think it's just a coincidence that result didn't differ much  :)

Reply:
No worries, if you want to fix it and push the change we can see if there is a big improvement. :+1:

Reply:
I have made the changes. Probably results are most consistent but sadly, there is no improvement in the runtimes :(

Reply:
Then probably something is not right in the code.  I also don't know what "no improvement" means, since you've already pointed out that the runtime is improved (but I don't know how much).  I'd suggest that we start debugging here, so some things to check:

 * How tight *should* the bound be?  Remember that the bound term is ||L_{t - 1} - L_t|| * (||x_a|| + ||x_b|| + 2 ||x_i||)---is that typically large or small?  Norm-centering the dataset can also help reduce the norms on average.

 * If the step size is sufficiently small, then || L_{t - 1} - L_t || should be so small that basically all evaluations should be bounded.  So you should be able to see, as the number of iterations increase (and the steps get smaller), that more and more pruning can happen.  If that's not happening, there's likely to be a bug.

 * The results may be different for different datasets.  So we should always be careful to run the benchmarks against a handful of datasets instead of just one or two, before saying that there is no improvement.

I think that in this case there is likely to be some bug though.  I'll take a look at the code when I have a chance.

Reply:
Just for the confirmation, have you read https://github.com/mlpack/mlpack/pull/1461#issuecomment-403009285 ?

Reply:
Yes, I read it, but at that time the code was only storing one `eval` for the entire dataset, which is definitely not right.  In fact I have found that the code never actually prunes anything.  Let me write up a longer comment shortly.